src/Microsoft.Extensions.Diagnostics.HealthChecks/HealthCheckService.cs

```diff
@@ -0,0 +1,118 @@
+        /// <remarks>
+        /// The key maps to the <see cref="IHealthCheck.Name"/> property of the health check, and the value is the <see cref="IHealthCheck"/>
+        /// instance itself.
+        /// </remarks>
+        public IReadOnlyDictionary<string, IHealthCheck> Checks { get; }
+
+        /// <summary>
+        /// Constructs a <see cref="HealthCheckService"/> from the provided collection of <see cref="IHealthCheck"/> instances.
+        /// </summary>
+        /// <param name="healthChecks">The <see cref="IHealthCheck"/> instances that have been registered in the application.</param>
+        public HealthCheckService(IEnumerable<IHealthCheck> healthChecks) : this(healthChecks, NullLogger<HealthCheckService>.Instance) { }
+
+        /// <summary>
+        /// Constructs a <see cref="HealthCheckService"/> from the provided collection of <see cref="IHealthCheck"/> instances, and the provided logger.
+        /// </summary>
+        /// <param name="healthChecks">The <see cref="IHealthCheck"/> instances that have been registered in the application.</param>
+        /// <param name="logger">A <see cref="ILogger{T}"/> that can be used to log events that occur during health check operations.</param>
+        public HealthCheckService(IEnumerable<IHealthCheck> healthChecks, ILogger<HealthCheckService> logger)
+        {
+            // Scan the list for duplicate names to provide a better error if there are duplicates.
+            var names = new HashSet<string>();
+            var duplicates = new List<string>();
+            foreach (var check in healthChecks)
+            {
+                if(!names.Add(check.Name))
+                {
+                    duplicates.Add(check.Name);
+                }
+            }
+
+            if(duplicates.Count > 0)
+            {
+                throw new InvalidOperationException($"Duplicate health checks were registered with the name(s): {string.Join(", ", duplicates)}");
+            }
+
+            Checks = healthChecks.ToDictionary(c => c.Name);
+            _logger = logger;
+
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                foreach (var check in Checks)
+                {
+                    _logger.LogDebug("Health check '{healthCheckName}' has been registered", check.Key);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Runs all the health checks in the application and returns the aggregated status.
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> which can be used to cancel the health checks</param>
+        /// <returns>
+        /// A <see cref="Task{T}"/> which will complete when all the health checks have been run,
+        /// yielding a <see cref="CompositeHealthCheckResult"/> containing the results.
+        /// </returns>
+        public async Task<CompositeHealthCheckResult> CheckHealthAsync(CancellationToken cancellationToken = default)
```

Comment:
Do we want to implement any concurrency throttling? E.g. having the service only allow one pass at a time via async lock. This may be related to your caching TODO. @glennc has already shared stories about DOS via heath checks.

Reply:
I like the idea of preventing simultaneous checks. I'll look in to that.